File: NeetCode/Linked_List/Find_The_Duplicate_Number/find__the__duplicate__number.py
from typing import Optional, List


class Solution:
    def findDuplicate(self, nums: List[int]) -> int:

        for index in range(len(nums)):
            if nums[abs(nums[index]) - 1] < 0:
                return abs(nums[index])
            nums[abs(nums[index]) - 1] *= -1

        return 0


if __name__ == "__main__":
    sol = Solution()
    nums = [[3, 1, 3, 4, 2], [1, 2, 3, 4, 4], [1, 2, 3, 2, 2], [1, 3, 4, 2, 2]]
    expected = [3, 4, 2, 2]

    for index, num in enumerate(nums):
        res = sol.findDuplicate(num)
        assert res == expected[index]

    # print(sol.findDuplicate(nums))
    print("Running Solution...")

File: NeetCode/Linked_List/Find_The_Duplicate_Number/Find_The_Duplicate_Number.cpp
// Find_The_Duplicate_Number.cpp
#include <algorithm>
#include <array>
#include <cassert>
#include <cctype>
#include <climits>
#include <cmath>
#include <deque>
#include <functional>
#include <iostream>
#include <map>
#include <numeric>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

using namespace std;

class Solution
{
public:
    int findDuplicate(vector<int> &nums)
    {
        for (int index = 0; index < nums.size(); index++)
        {
            int curr = nums[index];
            if (curr < 0)
            {
                return curr;
            }
            nums[abs(curr) - 1] *= 1;
        }

        return 0;
    }
};
